book/modular_code.md

```diff
@@ -0,0 +1,415 @@
+
+This approach helps you break complex logic down into small, understandable chunks that can be documented and tested more easily.
+
+When writing functions, it's also important to consider how they interact with other parts of your code. As a general rule of thumb, your code should run in the same way if a call to your function was replaced by the value that it would have returned. This is called 'referential transparency'.
+
+In practice, this means that your functions should not depend on or affect variables that have not been explicitly fed into them as arguments. For instance, a function should not add columns to a data table that has not been passed as an input to the function. Nor should the action of a function be affected by anything other than arguments that are passed to it. For example, running your function twice with the same inputs should always produce the same results.
+
+Avoiding such behaviours makes your code more transparent; it will be easier for users and developers to understand which functions affect which data without being concerned about hidden behaviours.
+In turn, this makes it easier to locate bugs in the code and assure its function by peer review.
+
+When it is not possible or practical to follow these practices, you should ensure that any 'side-effects' are adequately documented for both users and developers. This may be the case where your code interacts with a file, database or an external service. Ultimately, if you do signal where these kind of things might happen, someone trying to debug issues will know where to look.
+
+To summarise:
+
+- Make sure that functions are not overcomplicated; break down your code into smaller functions and build up your functionality with larger functions from these small building blocks.
+- Minimise the 'side-effects' of functions in order to make sure that your code is easy to debug and is transparent in its functionality.
+- Strive to make sure that running your function with the same inputs will produce the same results every time.
+
+### Group data and methods as classes
+
+Classes are fundamental parts of [object-orientated programming (OOP)](https://en.wikipedia.org/wiki/Object-oriented_programming). They create an association between data (attributes of the class) and logic (methods of the class). As you will see in the examples below, classes can be useful when representing real objects in our code.
+
+Although classes exist in R, [writing custom classeshttps://adv-r.hadley.nz/oo.html]() is less common than it is in Python (and other OOP enabling languages). Because of this, the following sub-section will focus primarily on Python classes. 
```

Comment:
I think this should fix the failing CI build
```suggestion
Although classes exist in R, [writing custom classes](https://adv-r.hadley.nz/oo.html) is less common than it is in Python (and other OOP enabling languages). Because of this, the following sub-section will focus primarily on Python classes. 
```